tutorials/basic-topics/sorting/bucket-sort.md

```diff
@@ -1,14 +1,315 @@
+    
+4. Concatenation:
+    
+    Concatenate all the sorted buckets in order to obtain the final sorted array.
+    
+
+We can understand the Bucket Sort by using Scatter-Gather Approach.
+
+![bucket sort example](https://github.com/Sanchita1304/leetcode-the-hard-way/assets/90080911/7045afe4-e20e-4019-acbf-b13dfcb7b886)
+
+
+In the above example, the unsorted array is given.
+
+- We have to create buckets with a range from 0 to 25.
+- The buckets range are 0-5, 5-10, 10-15, 15-20, 20-25.
+- After creating buckets we have to sort the buckets using different sorting algorithms of your own choice.
+- After sorting, gather the sorted elements from each bucket in order.
+
+## Complexity Analysis
+
+**Time Complexity : $$O(n^2)$$**, where n is the number of elements.
+
+**Space Complexity : $$O(n+k)$$**, where n is the number of elements and k is the number of buckets.
+
+
+## Example: [912 - Sort an Array](https://leetcode.com/problems/sort-an-array/description/)
+
+> An array of integers in the range [1,n] is given. We have to sort the array in ascending order and return it.
+
+There are two approaches we can use to solve the problem.
+
+### Approach 1: Using Uniform Bucket Sizes
+
+
+1. Find the Minimum and Maximum Values: 
+
+     The algorithm begins by finding the minimum and maximum values in the input array nums. This is done using the `min_element` and `max_element` functions from the C++ Standard Library. These values are crucial because they determine the range of values that need to be sorted.
+
+2. Create Buckets: 
+
+    Divides the range of input values into a series of "buckets." The number of buckets is determined by the range between the minimum and maximum values plus one `(bucket_range = max_val - min_val + 1)`. Each bucket is initially empty and will be used to count how many times each value appears in the input array.
+
+3. Distribute Elements into Buckets: 
+
+    The algorithm iterates through the input array nums. For each element, it calculates which bucket it belongs to by subtracting the min_val. It then increments the count in that bucket. This step effectively counts the occurrences of each value in the input array.
+
+4. Gather Sorted Elements:
+
+     After counting the occurrences of each value in the buckets, the algorithm reconstructs the sorted array. It iterates through the buckets in order and, for each non-empty bucket, adds the corresponding value back to the sorted array as many times as indicated by the count in the bucket. This process ensures that the values are added to the sorted array in ascending order.
+
+5. Return Sorted Array: 
+
+    Finally, the sorted array is returned as the result of the sortArray function.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+class Solution {
+public:
+    vector<int> sortArray(vector<int>& nums) {
+       // Find the minimum and maximum values in nums
```

Comment:
i would suggest to add the number as well. e.g. `// 1. Find the minimum and maximum values in nums`

(apply to all).